OrbitVulkanLayer/DispatchTable.h

```diff
@@ -0,0 +1,413 @@
+    dispatch_table.GetDeviceQueue2 = absl::bit_cast<PFN_vkGetDeviceQueue2>(
+        next_get_device_proc_addr_function(device, "vkGetDeviceQueue2"));
+
+    dispatch_table.CreateQueryPool = absl::bit_cast<PFN_vkCreateQueryPool>(
+        next_get_device_proc_addr_function(device, "vkCreateQueryPool"));
+    dispatch_table.ResetQueryPoolEXT = absl::bit_cast<PFN_vkResetQueryPoolEXT>(
+        next_get_device_proc_addr_function(device, "vkResetQueryPoolEXT"));
+
+    dispatch_table.CmdWriteTimestamp = absl::bit_cast<PFN_vkCmdWriteTimestamp>(
+        next_get_device_proc_addr_function(device, "vkCmdWriteTimestamp"));
+
+    dispatch_table.GetQueryPoolResults = absl::bit_cast<PFN_vkGetQueryPoolResults>(
+        next_get_device_proc_addr_function(device, "vkGetQueryPoolResults"));
+
+    dispatch_table.CmdBeginDebugUtilsLabelEXT = absl::bit_cast<PFN_vkCmdBeginDebugUtilsLabelEXT>(
+        next_get_device_proc_addr_function(device, "vkCmdBeginDebugUtilsLabelEXT"));
+    dispatch_table.CmdEndDebugUtilsLabelEXT = absl::bit_cast<PFN_vkCmdEndDebugUtilsLabelEXT>(
+        next_get_device_proc_addr_function(device, "vkCmdEndDebugUtilsLabelEXT"));
+
+    dispatch_table.CmdDebugMarkerBeginEXT = absl::bit_cast<PFN_vkCmdDebugMarkerBeginEXT>(
+        next_get_device_proc_addr_function(device, "vkCmdDebugMarkerBeginEXT"));
+    dispatch_table.CmdDebugMarkerEndEXT = absl::bit_cast<PFN_vkCmdDebugMarkerEndEXT>(
+        next_get_device_proc_addr_function(device, "vkCmdDebugMarkerEndEXT"));
+
+    void* key = GetDispatchTableKey(device);
+
+    {
+      absl::WriterMutexLock lock(&mutex_);
+      CHECK(!device_dispatch_table_.contains(key));
+      device_dispatch_table_[key] = dispatch_table;
+
+      CHECK(!device_supports_debug_utils_extension_.contains(key));
+      device_supports_debug_utils_extension_[key] =
+          dispatch_table.CmdBeginDebugUtilsLabelEXT != nullptr &&
+          dispatch_table.CmdEndDebugUtilsLabelEXT != nullptr;
+
+      CHECK(!device_supports_debug_marker_extension_.contains(key));
+      device_supports_debug_marker_extension_[key] =
+          dispatch_table.CmdDebugMarkerBeginEXT != nullptr &&
+          dispatch_table.CmdDebugMarkerEndEXT != nullptr;
+    }
+  }
+
+  void RemoveDeviceDispatchTable(VkDevice device) {
+    void* key = GetDispatchTableKey(device);
+    absl::WriterMutexLock lock(&mutex_);
+
+    CHECK(device_dispatch_table_.contains(key));
+    device_dispatch_table_.erase(key);
+
+    CHECK(device_supports_debug_utils_extension_.contains(key));
+    device_supports_debug_utils_extension_.erase(key);
+
+    CHECK(device_supports_debug_marker_extension_.contains(key));
+    device_supports_debug_marker_extension_.erase(key);
+  }
+
+  template <typename DispatchableType>
+  PFN_vkDestroyDevice DestroyDevice(const DispatchableType& dispatchable_object) {
+    absl::ReaderMutexLock lock(&mutex_);
+    void* key = GetDispatchTableKey(dispatchable_object);
```

Comment:
This is now under the lock but in previous methods it wasn't. What is the reason?

Reply:
There is no need to put this under a lock. I moved it upwards.